discovery/scaleway/instance.go

```diff
@@ -0,0 +1,169 @@
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scaleway
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/scaleway/scaleway-sdk-go/api/instance/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+const (
+	instanceLabelPrefix = metaLabelPrefix + "instance_"
+
+	instanceIDLabel         = instanceLabelPrefix + "id"
+	instanceIPIsPublicLabel = instanceLabelPrefix + "ip_is_public"
+	instanceIPVersionLabel  = instanceLabelPrefix + "ip_version"
+	instanceImageNameLabel  = instanceLabelPrefix + "image_name"
+	instanceNameLabel       = instanceLabelPrefix + "name"
+	instanceStateLabel      = instanceLabelPrefix + "status"
+	instanceTagsLabel       = instanceLabelPrefix + "tags"
+	instanceTypeLabel       = instanceLabelPrefix + "type"
+	instanceZoneLabel       = instanceLabelPrefix + "zone"
+)
+
+type instanceDiscovery struct {
+	*refresh.Discovery
+	Client    *scw.Client
+	Port      int
+	Zone      string
+	Project   string
+	AccessKey string
+	SecretKey string
+	Name      string
+	Tags      []string
+}
+
+func newInstanceDiscovery(conf *SDConfig) (*instanceDiscovery, error) {
+	d := &instanceDiscovery{
+		Port:      conf.Port,
+		Zone:      conf.Zone,
+		Project:   conf.Project,
+		AccessKey: conf.AccessKey,
+		SecretKey: string(conf.SecretKey),
+		Name:      conf.FilterName,
+		Tags:      conf.FilterTags,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "scaleway_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	profile, err := LoadProfile(conf)
+	if err != nil {
+		return nil, err
+	}
+	d.Client, err = scw.NewClient(
+		scw.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		scw.WithUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+		scw.WithProfile(profile),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up scaleway client: %w", err)
+	}
+
+	return d, nil
+}
+
+func (d *instanceDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := instance.NewAPI(d.Client)
+
+	req := &instance.ListServersRequest{
+		Zone: scw.Zone(d.Zone),
+	}
+
+	if d.Project != "" {
+		req.Project = scw.StringPtr(d.Project)
+	}
+
+	if d.Name != "" {
+		req.Name = scw.StringPtr(d.Name)
+	}
+
+	if d.Tags != nil {
+		req.Tags = d.Tags
+	}
+
+	servers, err := api.ListServers(req, scw.WithAllPages(), scw.WithContext(ctx))
```

Comment:
Everything above could be done in the newRefresher, WDYT?

Reply:
I would like to keep all that in the refresher to make it easy to understand what happens when we want to do a request on the API.